src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/CodeActions/Razor/CodeBlockService.cs

```diff
@@ -0,0 +1,151 @@
+            var indentedMethod = FormattingUtilities.AddIndentationToMethod(templateWithMethodName, options, startingIndent: 0);
+            var textWithCodeBlock = "@code {" + Environment.NewLine + indentedMethod + Environment.NewLine + "}";
+            var lastCharacterLocation = code.Source.Lines.GetLocation(code.Source.Length - 1);
+            var insertCharacterIndex = 0;
+            if (lastCharacterLocation.LineIndex == code.Source.Lines.Count - 1 && !IsLineEmpty(code.Source, lastCharacterLocation))
+            {
+                // The last line of the file is not empty so we need to place the code at the end of that line with a new line at the beginning.
+                insertCharacterIndex = lastCharacterLocation.CharacterIndex + 1;
+                textWithCodeBlock = $"{Environment.NewLine}{textWithCodeBlock}";
+            }
+
+            var eof = new Position(code.Source.Lines.Count - 1, insertCharacterIndex);
+            return new TextEdit()
+            {
+                Range = new Range { Start = eof, End = eof },
+                NewText = textWithCodeBlock
+            };
+        }
+
+        // A well-formed @code block exists, generate the method within it.
+        var openBraceLocation = openBrace.GetSourceLocation(code.Source);
+        var closeBraceLocation = closeBrace.GetSourceLocation(code.Source);
+        var previousLine = code.Source.Lines.GetLocation(closeBraceLocation.AbsoluteIndex - closeBraceLocation.CharacterIndex - 1);
+
+        var insertLineLocation =
+            openBraceLocation.LineIndex == closeBraceLocation.LineIndex || !IsLineEmpty(code.Source, previousLine)
+            ? closeBraceLocation
+            : previousLine;
+
+        var formattedGeneratedMethod = FormatMethodInCodeBlock(
+            code.Source,
+            csharpCodeBlock.GetSourceLocation(code.Source),
+            openBraceLocation.LineIndex,
+            closeBraceLocation.LineIndex,
+            insertLineLocation,
+            options,
+            templateWithMethodName);
+
+        var insertCharacter = openBraceLocation.LineIndex == closeBraceLocation.LineIndex
+            ? closeBraceLocation.CharacterIndex
+            : 0;
+        var insertPosition = new Position(insertLineLocation.LineIndex, insertCharacter);
+        return new TextEdit()
+        {
+            Range = new Range { Start = insertPosition, End = insertPosition },
+            NewText = formattedGeneratedMethod
+        };
+    }
+
+    private static string FormatMethodInCodeBlock(RazorSourceDocument source, SourceLocation codeBlockSourceLocation, int openBraceLineIndex, int closeBraceLineIndex, SourceLocation insertLocation, RazorLSPOptions options, string method)
+    {
+        var formattedGeneratedMethod = FormattingUtilities.AddIndentationToMethod(method, options, codeBlockSourceLocation.AbsoluteIndex - 5, codeBlockSourceLocation.CharacterIndex - 5, source);
```

Comment:
The -5 is for the `@code`, right? Is that assuming that the open brace is on the next line? I'm not quite sure what this is doing exactly. Perhaps a comment explaining it?

Reply:
Sure thing! I'll add a comment, but yes you are right the -5 is for the `@code`. The `codeBlockSourceLocation.AbsoluteIndex` and `codeBlockSourceLocation.CharacterIndex` will always point us to the end of the `@code` block regardless of where the open brace is. For formatting we want to know about the characters on the same line that come before the `@code`, hence the -5.